Find filters in map visualizations

jsoriano opened the pull request:
What
----

Support detection of filters in map visualizations. This detection is used by package-spec and elastic-package to check that dashboards in packages include filters.

How
----

Maps visualizations keep most of their definitions in the `mapStateJSON` object, including filters. This change adds the path to this object to the list of paths considered when looking for filters in `HasFilter()`.

Fix also the paths used to decode embedded JSON, so the fix mentioned also works for map panels whose content is encoded.

How to test
---

Export a dashboard containing a maps visualization stored by value, that includes a filter. `HasFilters()` for this visualization should return true.

Related issues
----
* In https://github.com/elastic/integrations/pull/9818 we needed to add an exception because elastic-package was not able to find the filters in a map visualization. elastic-package uses kbncontent under the hood for this check.

jsoriano commented on kbncontent.go:
```diff
@@ -28,7 +28,7 @@ func (v VisualizationDescriptor) findDocumentPathsAsString(paths []string) strin
 
 	for _, path := range paths {
-		if m.Get(path).IsStr() {
+		if m.Get(path).IsStr() && m.Get(path).Str() != "" {
```
Without this change this seems to return the first path it founds, even if it is empty. There are cases where the first path is empty, but there are others with values.

jsoriano commented on kbncontent.go:
```diff
@@ -178,10 +180,6 @@ func (v VisualizationDescriptor) TSVBType() string {
 
 // Title returns the visualization title
 func (v VisualizationDescriptor) Title() string {
-	if v.SavedObjectType != "visualization" {
-		return ""
-	}
-
```
Is there a reason to return a title only for visualizations?

jsoriano commented on kbncontent.go:
```diff
@@ -191,9 +189,14 @@ func (v VisualizationDescriptor) Title() string {
+		"embeddableConfig.attributes.kibanaSavedObjectMeta.searchSourceJSON",
+		"embeddableConfig.attributes.mapStateJSON",
+		"embeddableConfig.attributes.uiStateJSON",
+		"embeddableConfig.attributes.visState",
```
Adding the paths with the `embeddableConfig` prefix. Actually in all the test cases these paths are prefixed with `embeddableConfig`, so not sure if we need the others :thinking:

jsoriano commented on kbncontent.go:
```diff
@@ -207,6 +210,22 @@ func deserializeSubPaths(doc objx.Map) error {
 	}
 
+	jsonSlicePaths := []string{
+		"attributes.layerListJSON",
+		"embeddableConfig.attributes.layerListJSON",
```
`layerListJSON` is an slice, and fails to be decoded with `FromJSON`, we need to use `FromJSONSlice` for them.

jsoriano commented on testdata/dashboard.json:
```diff
@@ -944,6 +944,275 @@
+                "panelIndex": "d9771ad7-cec0-4e4b-a51e-bbc880a8af0d",
+                "title": "Unique IPs map encoded",
+                "type": "map",
+                "version": "8.10.2"
```
This one is a real example from the Apache package. I had to [add a validation exception](https://github.com/elastic/integrations/pull/9818/files#diff-f41b072e52719f3b4d8d80d9c2c7cfae9ddd44d78d03bf4d9e6381508a9ef79fR3) because we could not find a filter, but the filter is actually there.

The previous example is the same, but decoded. We don't support decoding maps yet in `elastic-package`, adding support in https://github.com/elastic/elastic-package/pull/1825

nreese commented:
Can you add some more background information in the description? What problem is this trying to solve? Where/how can we view the existing problem? How can we test the PR?   Thanks

jsoriano commented:
> Can you add some more background information in the description? What problem is this trying to solve? Where/how can we view the existing problem? How can we test the PR? Thanks

Added more info to the description. For testing I have added test cases to the unit test directly taken from real world examples. Let me know if I should prepare a build of  `elastic-package` including this fix for further manual testing.

Thanks!

nreese commented on kbncontent.go:
```diff
@@ -137,6 +137,7 @@ func (v VisualizationDescriptor) HasFilters() (bool, error) {
 	queryPaths := []string{
 		"attributes.kibanaSavedObjectMeta.searchSourceJSON.query.query",
 		"attributes.state.query.query",
+		"embeddableConfig.attributes.mapStateJSON.query.query",
```
Should you also add the path `embeddableConfig.attributes.mapStateJSON.filters`? 

Maps uses Kibana's unified search, which applies filtering from 2 locations. `query` captures the contents search bar while `filters` captures the contents of the filter bar. Both should be checked to know if a map panel applies a panel level filtering.

jsoriano replied:
I added it in https://github.com/elastic/kbncontent/pull/5/files#diff-922b6d6bcd00ced7f244a57230a1075d65e42bae8fe3e8c8ac18ce9b82d03d9cR154.

The structure for queries and filters are different, so there are two different checks, each one with its own list of paths.

nreese reviewed (approved):
kibana-presentation changes LGTM
code review only

mrodm reviewed (approved):
LGTM code changes
